### PR TITLE
Turn tests into a python module

### PR DESCRIPTION
**Pull request by hiker**
This is for #508. I am using relative imports in all of tests, with the exception of utilities_test.test_compile_str:
```
    test_compile = "psyclone.tests.psyclone_test_utils.Compile"
    monkeypatch.setattr(test_compile+".TEST_COMPILE", False)
    monkeypatch.setattr(test_compile+".TEST_COMPILE_OPENCL", False)
```
I could not get the monkey-patch to work with using a '.' (it does a split by '.' internally somewhere).

### CODE REVIEW

**Comment by hiker** [comment]
I believe the decrease in coverage of utilities_test is only because I replaced two lines (that are not used unless compilation is enabled) by three lines. So I think it is ready to go.

File: src/psyclone/tests/dynamo0p3_cma_test.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
 import pytest
 import fparser
 from fparser import api as fpapi
-from dynamo0p3_build import Dynamo0p3Build
+from psyclone.tests.dynamo0p3_build import Dynamo0p3Build
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/dynamo0p3_cma_test.py]: I did that before I learned about relative imports, and then forgot to change this.

File: src/psyclone/tests/dynamo0p3_transformations_test.py [hunk 2]
```diff
@@ -55,7 +55,8 @@
     Dynamo0p3AsyncHaloExchangeTrans, \
     Dynamo0p3KernelConstTrans
 from psyclone.configuration import Config
-from dynamo0p3_build import Dynamo0p3Build
+
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/dynamo0p3_transformations_test.py]: Done.

File: src/psyclone/tests/f2pygen_test.py [hunk 3]
```diff
@@ -42,7 +42,7 @@
     CallGen, AllocateGen, DeallocateGen, IfThenGen, DeclGen, TypeDeclGen,\
     CharDeclGen, ImplicitNoneGen, UseGen, DirectiveGen, AssignGen
 from psyclone.psyGen import InternalError
-from psyclone_test_utils import Compile, count_lines, line_number
+from .psyclone_test_utils import Compile, count_lines, line_number
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/f2pygen_test.py]: Renamed.

File: src/psyclone/tests/gocean1p0_build.py [hunk 4]
```diff
@@ -70,13 +70,10 @@ def __init__(self, tmpdir):
...
-        Returns the required flag to use the infrastructure library
-        dl_esm_inf for gocean1p0. Each parameter must be a separate
-        entry in the list, e.g.: ["-I", "/some/path"] and not ["-I
-        /some/path"].
-
-        :returns: A list of strings with the compiler flags required.
+        '''Returns the required flag to use the infrastructure library
+        dl_esm_inf for gocean1p0. Each parameter must be a separate entry
+        in the list, e.g.: ["-I", "/some/path"] and not ["-I /some/path"].
+        :returns: a list of strings with the compiler flags required.
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/gocean1p0_build.py]: Done

File: src/psyclone/tests/gocean1p0_transformations_test.py [hunk 5]
```diff
@@ -51,7 +51,9 @@
     GOceanOMPLoopTrans, KernelModuleInlineTrans, GOceanLoopFuseTrans, \
     ACCParallelTrans, ACCEnterDataTrans, ACCLoopTrans
 from psyclone.generator import GenerationError
-from psyclone_test_utils import count_lines, get_invoke, Compile
+from .psyclone_test_utils import count_lines, get_invoke, Compile
+
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/gocean1p0_transformations_test.py]: Done

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 6]
```diff
@@ -39,12 +39,12 @@
 from __future__ import absolute_import
 
 import pytest
-from fparser.two.parser import ParserFactory
 from fparser.common.readfortran import FortranStringReader
 from psyclone.psyir.backend.base import VisitorError
 from psyclone.psyir.backend.fortran import gen_intent, gen_dims, gen_kind, \
     FortranWriter
 from psyclone.psyGen import Symbol, Fparser2ASTProcessor, Node, CodeBlock
+from ...psyclone_test_utils import create_schedule
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Not sure about more robust ... depends what you want to do with the files :)   If you want to move all tests to be 'next' to src (which would then avoid the exception in the setup.py file, then relative imports are better. But am happy either way ... it's only a 'sed' statement to change them all :) So done.

File: src/psyclone/tests/psyclone_test_utils.py [hunk 7]
```diff
@@ -430,3 +438,29 @@ def get_invoke(algfile, api, idx=None, name=None):
...
+    :rtype: Subclass of :py:class:`psyclone.psyGen.Node`
+
+    '''
+    reader = FortranStringReader(code)
+    f2003_parser = ParserFactory().create(std="f2003")
+    parse_tree = f2003_parser(reader)
+
+    # Generate PSyIR schedule from fparser2 parse tree
+    processor = ast_processor()
+    schedule = processor.generate_schedule("tmp", parse_tree)
```
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/psyclone_test_utils.py]: Not sure what you are trying to tell me here ... do I need to change anything? Yes, I moved it from fortran_test to (now) utilities. Let me know if I need to do something.

**Comment by hiker** [comment]
All comments address.
I also decided to include a move of .../core/tests to .../tests/core. While I find it convenient to have tests in the same directory as the modules it is testing (makes for quick 'git diff','pytest', ... all from the same directory), it was inconsistent, and ... kind of fit with this PR 😀

File: src/psyclone/tests/psyclone_test_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hiker) [reply src/psyclone/tests/psyclone_test_utils.py]: Ah, got it now, and fixed.

File: src/psyclone/tests/f2pygen_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/f2pygen_test.py]: Great - thanks.

File: src/psyclone/tests/psyclone_test_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyclone_test_utils.py]: Sorry - it all made sense in my head ;-)

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: I take your point - I was only thinking of moving things around within the existing hierarchy. If we move the whole tests directory then we can tackle it then...

**Comment by arporter** [comment]
Thanks for making those changes and moving the `core` tests.
All tests pass or xfail (including with compilation) so functionally we are all good.
However...I thought about documentation and took a look at the user guide. That talks about `pip` installing the test dependencies which doesn't seem right given that pip does not install the test suite. Then again, it never has installed the test suite so maybe just add note saying that you'll need to install from source (rather than pypi) in order to run the suite? Once that's done this can be merged.

**Comment by hiker** [comment]
> However...I thought about documentation and took a look at the user guide. That talks about `pip` installing the test dependencies which doesn't seem right given that pip does not install the test suite. 
Then again, it never has installed the test suite so maybe just add note saying that you'll need to install from source (rather than pypi) in order to run the suite? Once that's done this can be merged.

OK, I have updated the docs, removed most references to testing from the user's guide (except a reference that running the test suite requires a source install). I've also updated the dev guide a bit, since I learned that setup.py supports things like "pip install .[test]" to install the dependencies. So I documented the latter and moved the dependencies into setup.py. I also made sure we consistently use 'pytest' (and not the old py.test).

**Pull request merged**